lib/Driver/Driver.cpp

```diff
@@ -1348,22 +1348,48 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
 
   if (const Arg *A = Args.getLastArg(options::OPT_g_Group)) {
     if (A->getOption().matches(options::OPT_g))
-      OI.DebugInfoKind = IRGenDebugInfoKind::Normal;
+      OI.DebugInfoLevel = IRGenDebugInfoLevel::Normal;
     else if (A->getOption().matches(options::OPT_gline_tables_only))
-      OI.DebugInfoKind = IRGenDebugInfoKind::LineTables;
+      OI.DebugInfoLevel = IRGenDebugInfoLevel::LineTables;
     else if (A->getOption().matches(options::OPT_gdwarf_types))
-      OI.DebugInfoKind = IRGenDebugInfoKind::DwarfTypes;
+      OI.DebugInfoLevel = IRGenDebugInfoLevel::DwarfTypes;
     else
       assert(A->getOption().matches(options::OPT_gnone) &&
              "unknown -g<kind> option");
   }
 
+  if (const Arg *A = Args.getLastArg(options::OPT_debug_info_format)) {
+    if (strcmp(A->getValue(), "dwarf") == 0)
+      OI.DebugInfoFormat = IRGenDebugInfoFormat::DWARF;
+    else if (strcmp(A->getValue(), "codeview") == 0)
+      OI.DebugInfoFormat = IRGenDebugInfoFormat::CodeView;
+    else
+      Diags.diagnose(SourceLoc(), diag::error_invalid_arg_value,
+                     A->getAsString(Args), A->getValue());
+  } else if (OI.DebugInfoLevel > IRGenDebugInfoLevel::None) {
+    // If -g was specified but not -debug-info-format, DWARF is assumed.
+    OI.DebugInfoFormat = IRGenDebugInfoFormat::DWARF;
+  }
+  if (OI.DebugInfoFormat > IRGenDebugInfoFormat::None &&
+      OI.DebugInfoLevel == IRGenDebugInfoLevel::None) {
+    // We have specified -debug-info-format but not -g.
+    const Arg *debugFormatArg = Args.getLastArg(options::OPT_debug_info_format);
+    Diags.diagnose(SourceLoc(), diag::error_option_missing_required_argument,
+                   debugFormatArg->getAsString(Args), "-g");
+  } else if (OI.DebugInfoFormat == IRGenDebugInfoFormat::CodeView &&
+             OI.DebugInfoLevel != IRGenDebugInfoLevel::Normal) {
```

Comment:
Ah, strictly speaking `-gnone -debug-info-format=codeview` should be allowed, right? (Could be useful for autogenerated command lines.)

Reply:
I can specifically allow `-g -debug-info-format=codeview` and `-gnone -debug-info-format=codeview`, but as the code is written that would also allow just `-dwarf-info-format=codeview`.

Reply:
Do we want any kind of warning in case someone forgets `-g`?

Reply:
I think what you had earlier is reasonable as long as it checks for the presence of a `-g*` flag rather than just "oh, the debug level is None".